adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -250,7 +256,23 @@ public static AuthenticationResult processUIResponseParams(Map<String, String> r
                     response.get(AuthenticationConstants.OAuth2.ERROR_CODES));
 
         } else if (response.containsKey(AuthenticationConstants.OAuth2.CODE)) {
+            // The header cloud_instance_host_name points to the right sovereign cloud to use for the given user
+            // Using this host name we construct the authority that will get the token request and we use this authority
+            // to save the token in the cache. The app should reinitialize AuthenticationContext with this authority for
+            // all subsequent requests.
             result = new AuthenticationResult(response.get(AuthenticationConstants.OAuth2.CODE));
+            final String cloudInstanceHostName = response.get(AuthenticationConstants.OAuth2.CLOUD_INSTANCE_HOST_NAME);
+            if (!StringExtensions.isNullOrBlank(cloudInstanceHostName)) {
+
+                final URL authorityUrl = StringExtensions.getUrl(mRequest.getAuthority());
+                final String newAuthorityUrlString = new Uri.Builder().scheme(HTTPS_PROTOCOL_STRING)
+                        .authority(cloudInstanceHostName)
+                        .path(authorityUrl.getPath())
+                        .build().toString();
```

Comment:
Why not complete the cloud_instance_name with 'https://login.' + $cloudInstanceHostName + '/common'? @addev-ashish